# Importing libraries

- pandas  for working with tables (DataFrames)
- numpy  for generating random data efficiently

In [2]:
import pandas as pd
import numpy as np

- Generating the number of rows

In [3]:
N = 500_000  

In [4]:
# Setting a seed so that results are reproducible
np.random.seed(42)

# Creating the dataset using a pandas DataFrame

In [5]:
data = pd.DataFrame({

    # Unique ID for each transaction
    "transaction_id": np.arange(1, N + 1),

    # Generate random dates within one year
    "date": pd.to_datetime("2023-01-01") + pd.to_timedelta(
        np.random.randint(0, 365, N), unit="D"
    ),

    # Randomly assign regions
    "region": np.random.choice(
        ["Central", "Eastern", "Western", "Northern"], N
    ),

    # Random product categories
    "product_category": np.random.choice(
        ["Food", "Electronics", "Clothing", "Furniture"], N
    ),

    # Quantity of items bought (1 to 9)
    "quantity": np.random.randint(1, 10, N),

    # Unit price between 5 and 500
    "unit_price": np.round(np.random.uniform(5, 500, N), 2),

    # Payment methods
    "payment_method": np.random.choice(
        ["Cash", "Mobile Money", "Card"], N
    ),

    # Customer type
    "customer_type": np.random.choice(
        ["Retail", "Wholesale"], N
    )
})


- Creating a derived column for total transaction amount


In [6]:
!pip3 install pyarrow


[notice] A new release of pip is available: 25.3 -> 26.0.1
[notice] To update, run: /Library/Frameworks/Python.framework/Versions/3.12/bin/python3.12 -m pip install --upgrade pip


In [7]:
# total_amount = quantity * unit_price
data["total_amount"] = data["quantity"] * data["unit_price"]

# Save the dataset as a CSV file (row-based storage)
data.to_csv("transactions.csv", index=False)

# Save the dataset as a Parquet file (column-based storage)
data.to_parquet("transactions.parquet")

print("Dataset created successfully!")

Dataset created successfully!


# File size comparison

In [9]:
import os
# Get file size of CSV in megabytes
csv_size_mb = os.path.getsize("transactions.csv") / (1024 * 1024)

# Get file size of Parquet in megabytes
parquet_size_mb = os.path.getsize("transactions.parquet") / (1024 * 1024)

print(f"CSV file size: {csv_size_mb:.2f} MB")
print(f"Parquet file size: {parquet_size_mb:.2f} MB")

CSV file size: 32.86 MB
Parquet file size: 7.03 MB


- The Parquet file is smaller than the CSV file because Parquet uses column-based storage and compression. CSV stores data as plain text, which is inefficient for large datasets. Big Data systems prefer columnar formats like Parquet because they reduce disk usage and improve query performance.

# Read performance Test

In [10]:
import time

In [12]:
# Loading the entire CSV file
start_time = time.time()
df_csv = pd.read_csv("transactions.csv")
csv_load_time = time.time() - start_time
print(f"CSV load time: {csv_load_time:.2f} seconds")

CSV load time: 1.09 seconds


In [13]:
#Loading the entire Parquet file
start_time = time.time()
df_parquet = pd.read_parquet("transactions.parquet")
parquet_load_time = time.time() - start_time
print(f"Parquet load time: {parquet_load_time:.2f} seconds")

Parquet load time: 1.51 seconds


In [14]:
#Loading only selected columns from Parquet
start_time = time.time()
df_parquet_cols = pd.read_parquet(
    "transactions.parquet",
    columns=["region", "total_amount"]
)
parquet_cols_load_time = time.time() - start_time

print(f"Parquet (2 columns) load time: {parquet_cols_load_time:.2f} seconds")

Parquet (2 columns) load time: 0.09 seconds


- Loading the Parquet file is faster than loading the CSV file because Parquet is optimized for analytical workloads. When only specific columns are selected, Parquet loads even faster and uses less memory. This demonstrates why Big Data systems avoid reading unnecessary data and instead scan only required columns.

# Results

- The CSV file size is: 32.86 MB
- The Parquet file size is: 7.03 MB
- The CSV file load time is: 1.09 seconds
- The Parquet file load time: 1.51 seconds
- The Parquet (2 columns) file load time is: 0.09 seconds

The Parquet file was significantly smaller than the CSV file (7.03 MB vs 32.86 MB). This is because Parquet uses column-based storage and compression, while CSV stores data as plain text. Big Data systems prefer Parquet to reduce storage costs and improve I/O efficiency.

Although the CSV file loaded slightly faster than the Parquet file, this is expected for moderate-sized datasets due to Parquet’s decompression overhead. However, when only two columns were selected, Parquet loaded significantly faster. This demonstrates the advantage of columnar storage in analytical workloads

# Memory Usage

- The memory used before was 90MB's loading the parquet file thus moderate
- The memory used after by the RAM was 127MB's thus high

- Memory usage increased from approximately 90 MB to 127 MB when loading the full dataset, indicating higher RAM consumption during full data scans. Column selection from Parquet reduced memory usage, highlighting why Big Data systems avoid reading unnecessary columns.

# Big data problem on a small machine

- Defining the number of rows to read at a time this is aimed at reducing memory usage 

In [15]:
CHUNK_SIZE = 50_000

- Creating the total number of records

In [16]:
# Initializing a counter for total records
total_records = 0

# Reading the CSV file in chunks
for chunk in pd.read_csv("transactions.csv", chunksize=CHUNK_SIZE):
    
    # Adding the number of rows in the current chunk
    total_records += len(chunk)
print("Total number of records:", total_records)

Total number of records: 500000


- The total number of records was calculated by reading the CSV file in chunks instead of loading the entire dataset into memory. This approach prevents memory overload and is suitable for large datasets.

# Average Transaction Value per Product category

In [17]:
# Dictionaries to store cumulative sums and counts
total_amount_sum = {}
transaction_count = {}

# Process the CSV file chunk by chunk
for chunk in pd.read_csv("transactions.csv", chunksize=CHUNK_SIZE):
    
    # Group data by product category within the chunk
    grouped = chunk.groupby("product_category")["total_amount"].agg(["sum", "count"])
    
    # Accumulate results across all chunks
    for category, row in grouped.iterrows():
        total_amount_sum[category] = total_amount_sum.get(category, 0) + row["sum"]
        transaction_count[category] = transaction_count.get(category, 0) + row["count"]

# Calculate the average transaction value per category
average_transaction_value = {
    category: total_amount_sum[category] / transaction_count[category]
    for category in total_amount_sum
}

print("Average transaction value per category:")
print(average_transaction_value)

Average transaction value per category:
{'Clothing': 1259.08098429613, 'Electronics': 1262.440670266557, 'Food': 1260.870070498439, 'Furniture': 1260.529748216433}
